src/ripple/consensus/Validations.h

```diff
@@ -0,0 +1,506 @@
+    struct LedgerID;
+    struct NodeID;
+    struct NodeKey;
+
+    struct Validation
+    {
+        // Ledger ID associated with this validation
+        LedgerID ledgerID() const;
+
+        // Optional sequence number of validation's ledger
+        boost::optional<std::size_t> seq() const
+
+        // When the validation was signed
+        NetClock::time_point signTime() const;
+
+        // When the validation was first observed by this node
+        NetClock::time_point seenTime() const;
+
+        // Signing key of node that published the validation
+        NodeKey key() const;
+
+        // Identifier of node that published the validaton
+        NodeID nodeID() const;
+
+        // Whether the publishing node was trusted at the time the validation
+        // arrived
+        bool trusted() const;
+
+        // Set the previous validation ledger from this publishing node that this
+        // validation replaced
+        void setPreviousLedgerID(LedgerID &);
+
+        // Check if this validation had the given ledger ID as its prior ledger
+        bool isPreviousLedgerID(LedgerID const& ) const;
+
+        // ... implementation specific
+    };
+
+    class Derived : public Validations<Derived, Validation>
+    {
+        Derived(ValidationParms const & p, clock_type & c, ...)
+            : Validations(p, c), ...
+
+        // Handle a newly stale validation
+        void onStale(Validations && );
+
+        // ... implementation specific
+    };
+    @endcode
+
+    @tparam Derived Provides functions conforming to the CRTP interface
+    @tparam Validation Conforming type representing a ledger validation
+
+*/
+template <class Derived, class Validation>
+class Validations
+{
+    template <typename T>
+    using decay_result_t = std::decay_t<std::result_of_t<T>>;
+
+    using LedgerID =
+        decay_result_t<decltype (&Validation::ledgerID)(Validation)>;
+    using NodeKey = decay_result_t<decltype (&Validation::key)(Validation)>;
+    using NodeID = decay_result_t<decltype (&Validation::nodeID)(Validation)>;
+    using ValidationMap = hash_map<NodeKey, Validation>;
+
+    //! The latest validation from each node
+    ValidationMap current_;
+
+    //! Recent validations from nodes, indexed by ledger identifier
+    beast::aged_unordered_map<
+        LedgerID,
+        ValidationMap,
+        std::chrono::steady_clock,
+        beast::uhash<>>
+        byLedger_;
+
+    //! Parameters to determine validation staleness
+    ValidationParms parms_;
+
+    //! @return The Derived class that implements the CRTP requirements.
+    Derived&
+    impl()
+    {
+        return static_cast<Derived&>(*this);
+    }
+
+protected:
+    /** Iterate current validations.
+
+        Iterate current validations, optionally removing any stale validations
+        if a time is specified
+
+        @param t (Optional) Time used to determine staleness
+        @param f Callable with signature (NodeKey const &, Validations const &)
```

Comment:
It had come up in the past, it seems to be a bug in how docca renders some templates.